File: jdk6u21_src/hotspot/agent/src/share/classes/sun/jvm/hotspot/jdi/StringReferenceImpl.java
/*
 * Copyright (c) 2002, 2003, Oracle and/or its affiliates. All rights reserved.
 * ORACLE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 */

package sun.jvm.hotspot.jdi;

import com.sun.jdi.*;
import sun.jvm.hotspot.oops.Instance;
import sun.jvm.hotspot.oops.OopUtilities;

public class StringReferenceImpl extends ObjectReferenceImpl
    implements StringReference
{
    private String value;

    StringReferenceImpl(VirtualMachine aVm, sun.jvm.hotspot.oops.Instance oRef) {
        super(aVm,oRef);
        value = OopUtilities.stringOopToString(oRef);
    }

    public String value() {
        return value;
    }

    public String toString() {
        return "\"" + value() + "\"";
    }
}

File: trunk/jdk6u21_src/j2se/src/share/back/StringReferenceImpl.c
/*
 * @(#)StringReferenceImpl.c	1.21 10/03/23
 *
 * Copyright (c) 2006, Oracle and/or its affiliates. All rights reserved.
 * ORACLE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 */

#include "util.h"
#include "StringReferenceImpl.h"
#include "inStream.h"
#include "outStream.h"

static jboolean 
value(PacketInputStream *in, PacketOutputStream *out)
{
    JNIEnv *env;
    jstring string;
    
    env = getEnv();
    
    string = inStream_readStringRef(env, in);
    if (inStream_error(in)) {
        return JNI_TRUE;
    }

    WITH_LOCAL_REFS(env, 1) {

        char *utf;
        
        utf = (char *)JNI_FUNC_PTR(env,GetStringUTFChars)(env, string, NULL);
        (void)outStream_writeString(out, utf);
        JNI_FUNC_PTR(env,ReleaseStringUTFChars)(env, string, utf);

    } END_WITH_LOCAL_REFS(env);
    
    return JNI_TRUE;
}

void *StringReference_Cmds[] = { (void *)0x1
    ,(void *)value};
